### DISCUSSION TITLE
Is it possible to change the gamepad name in Control Panel/Ingame?

### DISCUSSION THREAD

**Discussion by WallK** [afpineda/OpenSourceSimWheelESP32#10] [Q&A] [answered]
Hello!
The project supports renaming the device and manufacturer
But the "Controller" name stays at "4 axis 128 buttons..." for me
![2024-07-19_05-00-43](https://github.com/user-attachments/assets/0ad22c72-a27e-426c-ba19-aa73bf3a62a5)
Is it possible to rename it?
Thanks!

**Comment by afpineda** [comment, marked as answer]
Hi, WallK.
In short, the answer is "no, but it should".
I'm aware of that and I would like to change the controller name, too. 
But I found no way to do so in the underlying API (NimBLE or USB). 
This issue is related to "USB string descriptors". That's all I know.
If you have any clue, please, let me know.

**Comment by afpineda** [comment]
I found this post which explains how to rename devices in Windows:
[https://www.eightforums.com/threads/tutorial-how-to-change-device-names-in-device-manager.15321/](https://www.eightforums.com/threads/tutorial-how-to-change-device-names-in-device-manager.15321/)

May work as a workaround.

**Comment by WallK** [comment]
I think this will rename all the ESP32 wheel devices at the same time
I will try this, thanks
Looking around how to implement it on the firmware level too

**Comment by simachines** [comment]
https://github.com/lemmingDev/ESP32-BLE-Gamepad/issues/35

**Reply by afpineda** [reply]
> [lemmingDev/ESP32-BLE-Gamepad#35](https://github.com/lemmingDev/ESP32-BLE-Gamepad/issues/35)

As this source suggest, the proper name should be set in the advertising data, as attribute ID "0x0100".
I did some test, but found no way to do so in *NimBLE-Arduino*.

**Comment by afpineda** [comment]
Hi, Wallk.

I found a proper way to rename HID devices in Windows:

1. Open the "registry editor" (regedit.exe) **with user privileges** (not administrator).
2. Navigate to the following folder (registry key):

   > KEY_CURRENT_USER\System\CurrentControlSet\Control\MediaProperties\PrivateProperties\Joystick\OEM

3. Inside that folder, locate the subfolder **"VID_1D50&PID_FFFF"** (assuming your device uses BLE connectivity).
4. Double-click (edit) the value "**OEMName**".
5. Set a custom display name and hit "enter".

Next project version will include explicit instructions for this.

**Reply by WallK** [reply]
> Hi, Wallk.
> 
> I found a proper way to rename HID devices in Windows:
> 
> 1. Open the "registry editor" (regedit.exe) **with user privileges** (not administrator).
> 2. Navigate to the following folder (registry key):
>    > KEY_CURRENT_USER\System\CurrentControlSet\Control\MediaProperties\PrivateProperties\Joystick\OEM
> 3. Inside that folder, locate the subfolder **"VID_1D50&PID_FFFF"** (assuming your device uses BLE connectivity).
> 4. Double-click (edit) the value "**OEMName**".
> 5. Set a custom display name and hit "enter".
> 
> Next project version will include explicit instructions for this.

That doesn't make a lot of sense
This will rename all the boards that are sharing this VID and PID (for example, if I have two button boxes)
Also, it will be back to nothing in another PC

**Reply by afpineda** [reply]
> That doesn't make a lot of sense 

Indeed. From my investigations, this seems to be Microsoft's fault on BLE HID devices. 
USB and Bluetooth classic work as expected.

> This will rename all the boards that are sharing this VID and PID (for example, if I have two button boxes) 

Yes, but you may set set another PID for your second device, as described [here](https://github.com/afpineda/OpenSourceSimWheelESP32/blob/main/doc/RenameDeviceWin_en.md)

> Also, it will be back to nothing in another PC

Yes. Even worst, it will be back to nothing in another user account. This is just a workaround. 

You may export the registry key (right-click > export) for other computers or for backup. 
Then, a double-click on the ".reg" file do all the work.

**Reply by WallK** [reply]
That's really messed up
How does the regular devices deal with it?
Do they have access to something we don't?

**Reply by afpineda** [reply]
> That's really messed up How does the regular devices deal with it? Do they have access to something we don't?

No Idea. Maybe they use [INF files](https://learn.microsoft.com/en-us/windows-hardware/drivers/install/using-an-extension-inf-file). But those need to be signed by Microsoft.

**Reply by WallK** [reply]
> No Idea. Maybe they use [INF files](https://learn.microsoft.com/en-us/windows-hardware/drivers/install/using-an-extension-inf-file). But those need to be signed by Microsoft.

I don't think this is the way for some chinese manufacturers
For example, random cheap device I've got was designed way after the last update to my system
And it's still has the proper name
I think we are missing something here

**Reply by afpineda** [reply]
The API has been changed as of version 4.2.0 to enable a configurable PID for each individual device.

**Comment by simachines** [comment]
A simple solution to this is connecting USB HID which tells windows the name for its pid/vid and when you connect BLE HID device with same PID/VID windows will think it's exactly the same device and will appear the way you want it to. I am using both usb hid and ble hid on my esp32 s3 so it works quite well. All that user needs to do is plug the usb in once and from that point the ble hid will always have the right name.

**Reply by simachines** [reply]
A benefit of that is you can use either connection type and you won't lose your button bindings in games

**Reply by afpineda** [reply]
> A simple solution to this is connecting USB HID which tells windows the name for its pid/vid and when you connect BLE HID device with same PID/VID windows will think it's exactly the same device and will appear the way you want it to. 

Maybe this is the trick used by some chinesse manufacturers, but hardware identifiers are **licensed**. So, there are legal concerns. 
In our context, any VID/PID you choose must be constrained to your personal sphere (this includes the default VID/PID in this project). See [HID notes](https://github.com/afpineda/OpenSourceSimWheelESP32/blob/main/doc/firmware/HID_notes.md).

**Reply by afpineda** [reply]
> Maybe this is the trick used by some chinesse manufacturers, but hardware identifiers are **licensed**. So, there are legal concerns. 

I have some fresh information here.

I have a gamepad manufactured by MSI (a well-known manufacturer). I checked the VID/PID for this device: 045E/028E.
According to [this database](https://devicehunt.com/view/type/usb/vendor/045E/device/028E), that pair is licensed to Microsoft for the XBOX 360 Controller. That explains why the display name for this device is "Controller (XBOX 360 for Windows)", not "MSI controller" or something like that.

This is a conjeture: Microsoft licensed some VID/PID pairs for generic devices, and they allow other manufactures to use them as long as there is no driver conflict.

**Reply by simachines** [reply]
Have you tried setting the vid/pid the same as generic controller to see if the name appears. You can try a xbox controller on https://hardwaretester.com/gamepad it will show the name unkown if its not reporting a name regardless if its showing up correctly in windows and you can try the same thing with micro controller to see if there is a difference. In regards to legal concerns they cant really do anything about it if you use it on your own pc. You can also apply for opensource vid/pid.

**Reply by afpineda** [reply]
> Have you tried setting the vid/pid the same as generic controller to see if the name appears. 

Yes. It works in BLE devices. USB devices make big trouble.

**Comment by afpineda** [comment]
As of version 4.3.0, you have the option to change the VID/PID (BLE only) and the display name (BLE and Windows, only) from the companion app.

I did some test. Windows gets crazy if you change the VID/PID on USB devices.

**Reply by simachines** [reply]
I have not had issue with chaning the pid vid of usb devices. It should simply appear as a different device. What did you see it do?

**Reply by afpineda** [reply]
> I have not had issue with chaning the pid vid of usb devices. It should simply appear as a different device. What did you see it do?

I put the VID/PID of the XBOX 360 controller into my USB device. Then it was not recognized as a HID device anymore (nor any other kind).

Note that VID/PID is inherent to all USB devices (not just USB HID devices), in contrast to BLE. Windows makes advantage of that data to determine which driver to load. You could fool Windows into loading a driver not matching your device.

In contrast, BLE drivers are divided into generic "classes" since there is no unique device identification. In BLE HID devices, PID/VID is just a service attribute in the HID over GATT specification to make them compatible with the original USB HID specification.

**Reply by simachines** [reply]
Xbox controller also needs to be compatible with x input i think. Try the ble hid composite library xinput example.
